bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/DataLayerFactory.java

```diff
@@ -0,0 +1,144 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2020 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal;
+
+import com.adobe.cq.wcm.core.components.models.DataLayerProvider;
+import com.day.cq.dam.api.Asset;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ValueMap;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+
+public class DataLayerFactory {
+
+    static public Map<String, Object> build(DataLayerProvider provider) {
+        MapWrapper data = new MapWrapper();
+        data.put("id", invoke(provider, "getDataLayerId"));
+        data.put("type", invoke(provider, "getDataLayerType"));
+        data.put("name", invoke(provider, "getDataLayerName"));
+        data.put("path", invoke(provider, "getDataLayerPath"));
+        data.put("title", invoke(provider, "getDataLayerTitle"));
+        data.put("template", invoke(provider, "getDataLayerTemplate"));
+        // data.put("src", invoke(provider, "getDataLayerSrc"));
+        data.put("text", invoke(provider, "getDataLayerText"));
+        data.put("tags", invoke(provider, "getDataLayerTags"));
+        data.put("asset", getAssetMetadata(provider));
+        data.put("linkUrl", invoke(provider, "getDataLayerLinkUrl"));
+        data.put("language", invoke(provider, "getDataLayerLanguage"));
+        data.put("itemsCount", invoke(provider, "getDataLayerItemsCount"));
+        data.put("activeItem", invoke(provider, "getDataLayerActiveItem"));
+        data.put("expandedItems", getExpandedItems(provider));
+        return data.getMap();
+    }
+
+    static private Object invoke(DataLayerProvider provider, String method) {
+        try {
+            Method providerMethod = provider.getClass().getMethod(method);
```

Comment:
Check for provider and method to be null?

Reply:
I'll add a check for null provider. If the method doesn't exist, NoSuchMethodException is handled in the catch block